### DISCUSSION TITLE
Sharing ownership between Python and C++

### DISCUSSION THREAD

**Discussion by eliotheinrich** [wjakob/nanobind#378] [Q&A]
I would like to share ownership of, e.g., and integer between Python and C++, looking something like this:
```
#include <nanobind/nanobind.h>
#include <nanobind/stl/shared_ptr.h>

class TestClass {
    public:
        std::shared_ptr<int> i;

        TestClass(std::shared_ptr<int> i) : i(i) {}
};

NB_MODULE(test_module, m) {
    nanobind::class_<TestClass>(m, "TestClass")
        .def(nanobind::init<std::shared_ptr<int>>());
}
```
When I compile, I run into an error:
```
.../site-packages/nanobind/include/nanobind/stl/shared_ptr.h:67:19: error: static assertion failed: Conversion of ``shared_ptr<T>`` requires that ``T`` is handled by nanobind's regular class binding mechanism. However, a type caster was registered to intercept this particular type, which is not allowed.
     static_assert(is_base_caster_v<Caster>,
                   ^~~~~~~~~~~~~~~~~~~~~~~~
```
I am relatively new to nanobind and I haven't found any answers about how to do this in the documentation; any clarification is appreciated!

**Comment by wjakob** [comment]
As the message says, it's not supported. You can only use the builtin ``shared_ptr`` type caster support for things that can be bound via `nb::class_<...>`. Builtin arithmetic types like `int` aren't in that category.
